compiler/rustc_interface/src/queries.rs

```diff
@@ -156,13 +156,11 @@ impl<'tcx> Queries<'tcx> {
 
     pub fn crate_name(&self) -> Result<&Query<String>> {
         self.crate_name.compute(|| {
-            Ok(match self.compiler.crate_name {
-                Some(ref crate_name) => crate_name.clone(),
-                None => {
-                    let parse_result = self.parse()?;
-                    let krate = parse_result.peek();
-                    find_crate_name(self.session(), &krate.attrs, &self.compiler.input)
-                }
+            Ok({
+                let parse_result = self.parse()?;
+                let krate = parse_result.peek();
+                // parse `#[crate_name]` even if `--crate-name` was passed, to make sure it matches.
+                find_crate_name(self.session(), &krate.attrs, &self.compiler.input)
```

Comment:
Hm, I am a bit worried -- is this a breaking change? It sounds like it would be, since we presumably weren't validating previously?

Reply:
See https://github.com/rust-lang/rust/pull/79586#issuecomment-736211297 - this is the same behavior as before, just easier to understand. Previously, the crate name would never be Some so this would always run.